File: DecodeString.py

# Version 3, recrusive
class Solution:
    def decodeString(self, s: str) -> str:
        return self.dfs(s, 0)
    def dfs(self, s, i):
        result = ""
        number = 0
        while i < len(s):
            if s[i].isdigit():
                number = number * 10 + int(s[i])
            elif s[i].isalpha():
                result += s[i]
            elif s[i] == '[':
                i, tmp = self.dfs(s, i + 1)
                result += tmp * number
                number = 0
            else:
                return i, result
            i += 1
        return result

        
# Version 2, stack
# class Solution:
#     def decodeString(self, s: str) -> str:
#         stack = []
#         num = 0
#         string = ""
#         for char in s:
#             if char.isdigit():
#                 num = 10 * num + int(char)
#             elif char.isalpha():
#                 string += char
#             elif char == '[':
#                 stack.append((string, num))
#                 num = 0
#                 string = ""
#             else:
#                 last_str, last_num = stack.pop(-1)
#                 # string += last_str * last_num
#                 string = last_str + string * last_num
#         return string


# Version 1, stack
# class Solution:
#     def decodeString(self, s: str) -> str:
#         stack = []
#         # print(s)
#         for char in s:
#             if char != ']':
#                 stack.append(char)
#             else:
#                 sub_string = ""
#                 while len(stack) > 0 and stack[-1] != '[':
#                     sub_char = stack[-1]
#                     stack.pop(-1)
#                     sub_string = sub_char + sub_string
#                 if len(stack) > 0:
#                     stack.pop(-1) # pop '['
#                 # deal with number
#                 number = ""
#                 while len(stack) > 0 and stack[-1] in [str(i) for i in range(10)]:
#                     sub_char = stack[-1]
#                     stack.pop(-1)
#                     number = sub_char + number
#                 # print(number)
#                 number = int(number)
#                 for i in range(number):
#                     stack.append(sub_string)
#         final_string = ""
#         while len(stack) > 0:
#             sub_string = stack[-1]
#             stack.pop(-1)
#             final_string = sub_string + final_string
#         return final_string

if __name__ == "__main__":

    solution = Solution()
    # s = "3[a]2[bc]"
    s = "abc3[cd]xyz"
    decode_s = solution.decodeString(s)
    print(decode_s)

File: DecodeString.cpp
//
// Created by Golfer on 2020/7/28.
//

#include <string>
#include <stack>
using namespace std;
#include <cstdio>
#include <iostream>

// Version 1, Stack
//class Solution {
//public:
//    string decodeString(string s) {
//        stack<string> stack_string;
//        stack<int> stack_num;
//        int number = 0;
//        string result= "";
//        for (char &c : s) {
////            if ('0' <= c <= '9') { // wrong !!!
////                number = number * 10 + c - '0';
////            }
//            if ('0' <= c && c <= '9') {
//                number = number * 10 + c - '0';
//            }
//            else if (c == '[') {
//                stack_num.push(number);
//                number = 0;
//                stack_string.push(result);
//                result = "";
//            }
//            else if (c == ']') {
//                int num = stack_num.top();
//                stack_num.pop();
//                string last_string = stack_string.top();
//                stack_string.pop();
//                string result_tmp = result;
//                for (int i = 0; i < num - 1; i++) {
//                    result += result_tmp;
//                }
//                result = last_string + result;
////                result = last_string + result
//            }
//            else {
//                result += c;
//            }
//        }
//        return result;
//    }
//};

// Version 2, recrusive
class Solution {
public:
    string decodeString(string s) {
        return dfs(s, 0).second;
    }
    pair<int, string> dfs(string s, int i) {
        int number = 0;
        string result = "";
        while (i < s.size()) {
            char c = s[i];
            if (c >= '0' && c <= '9') {
                number = number * 10 + c - '0';
            }
            else if (c == '[') {
                pair<int, string> dfs_call = dfs(s, i + 1);
                i = dfs_call.first;
                string tmp = dfs_call.second;
                for (int j = 0; j < number; j++)
                    result += tmp;
                number = 0;
            }
            else if (c == ']') {
                return make_pair(i, result);
            }
            else {
                result += c;
            }
            ++i;
        }
        return make_pair(-1, result);
    }
};

int main() {
    string s = "3[a]2[bc]";
    Solution solution = Solution();
    string decode_s = solution.decodeString(s);
    cout << decode_s << endl;
    return 0;
}

